cmd/frontend/graphqlbackend/slow_requests_tracer.go

```diff
@@ -0,0 +1,249 @@
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// slowRequestRedisFIFOListDefaultSize sets a default value for the FIFO list.
+const slowRequestRedisFIFOListDefaultSize = 5000
+
+// slowRequestRedisFIFOListPerPage sets the default count of returned request.
+const slowRequestRedisFIFOListPerPage = 50
+
+// slowRequestRedisFIFOList is a FIFO redis cache to store the slow requests.
+var slowRequestRedisFIFOList = rcache.NewFIFOList("slow-graphql-requests-list", slowRequestRedisFIFOListDefaultSize)
+
+// slowRequestConfWatchOnce enables to ensure we're not watching for conf updates more than once.
+var slowRequestConfWatchOnce sync.Once
+
+// captureSlowRequest stores in a redis cache slow GraphQL requests.
+func captureSlowRequest(ctx context.Context, logger log.Logger, req *types.SlowRequest) {
+	slowRequestConfWatchOnce.Do(func() {
+		conf.Watch(func() {
+			limit := conf.Get().ObservabilityCaptureSlowGraphQLRequestsLimit
+			if limit == 0 {
+				limit = slowRequestRedisFIFOListDefaultSize
+			}
+			slowRequestRedisFIFOList = rcache.NewFIFOList("slow-graphql-requests-list", limit)
+		})
+	})
+
+	b, err := json.Marshal(req)
+	if err != nil {
+		logger.Warn("failed to marshal slowRequest", log.Error(err))
+		return
+	}
+	if err := slowRequestRedisFIFOList.Insert(b); err != nil {
+		logger.Warn("failed to capture slowRequest", log.Error(err))
+	}
+}
+
+// getSlowRequestsAfter returns the last limit slow requests, starting at the request whose ID is set to after.
+func getSlowRequestsAfter(ctx context.Context, list *rcache.FIFOList, after int, limit int) ([]*types.SlowRequest, error) {
+	raws, err := list.Slice(ctx, after, after+limit-1)
+	if err != nil {
+		return nil, err
+	}
+
+	reqs := make([]*types.SlowRequest, 0, len(raws))
+	for i, raw := range raws {
+		var req types.SlowRequest
+		if err := json.Unmarshal(raw, &req); err != nil {
+			return nil, err
+		}
+		req.ID = strconv.Itoa(i + after)
+		reqs = append(reqs, &req)
+	}
+	return reqs, nil
+}
+
+// SlowRequests returns a connection to fetch slow requests.
+func (r *schemaResolver) SlowRequests(ctx context.Context, args *slowRequestsArgs) (*slowRequestConnectionResolver, error) {
+	if conf.Get().ObservabilityCaptureSlowGraphQLRequestsLimit == 0 {
+		return nil, errors.New("slow graphql requests capture is not enabled")
+	}
+	// 🚨 SECURITY: Only site admins may list outbound requests.
+	if err := auth.CheckCurrentUserIsSiteAdmin(ctx, r.db); err != nil {
+		return nil, err
+	}
+	var after string
+	if args.After != nil {
+		err := relay.UnmarshalSpec(*args.After, &after)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		after = "0"
+	}
+	return &slowRequestConnectionResolver{
+		after:   after,
+		perPage: slowRequestRedisFIFOListPerPage,
+	}, nil
+}
+
+type slowRequestConnectionResolver struct {
+	after string
+
+	err     error
+	once    sync.Once
+	reqs    []*types.SlowRequest
+	perPage int
```

Comment:
Nit: I'd group them semantically

```suggestion
	after string
	perPage int

	err     error
	once    sync.Once
	reqs    []*types.SlowRequest
```